src/query.rs

```diff
@@ -1113,6 +1131,14 @@ impl<'q, Q: Query> PreparedQueryBorrow<'q, Q> {
 
         unsafe { PreparedQueryIter::new(self.meta, self.archetypes, state.iter()) }
     }
+
+    /// Provides random access to the results of the prepared query
+    pub fn map<'i>(&'i mut self) -> PreparedQueryMap<'i, Q> {
```

Comment:
This is currently only defined for `PreparedQuery` as this would allow borrowing the `Vec<Option<Q::Fetch>>` within `PreparedQueryMap` to amortize its allocation costs.

I am not sure what the preferred options for hecs are:
* Should I make the current implementation of `PreparedQueryMap` available as `QueryMap` based off of `Query`?
* Should I put a `Box<[Q::Fetch]>` into `PreparedQuery` to amortize the allocation costs but making `PreparedQuery` larger unconditionally?

Reply:
I don't have any problem with making `PreparedQuery` slightly larger, though amortizing that allocation may be a premature opt since it's off the critical hot path of `get`.

I would like to support non-prepared queries as well for ergonomics. If it's desirable to borrow from `PreparedQuery`, perhaps `Cow` could be used to handle both cases gracefully in a single type?

Reply:
> I would like to support non-prepared queries as well for ergonomics. If it's desirable to borrow from PreparedQuery, perhaps Cow could be used to handle both cases gracefully in a single type?

I tried that but it does not really help as it is not just how the `Fetch` instances are stored but also that the states need to be prepared instead of just iterated. I will try adding separate `View` and `PreparedView` types instead.

Reply:
Scratch that, I did make the structure generic over how the fetch instances are stored with two different constructor methods `new` and `new_prepared`.

I went for `AsRef<[Option<Q::Fetch>]>` instead of `Cow<'q, [Option<Q::Fetch>]>` though as the dispatch over the `Cow`'s discriminant would end up in the fast path for `get(_mut)` even though the type of storage is fixed after construction. This does have the downside of exposing the storage in the return type of the public methods though.

Reply:
> the dispatch over the Cow's discriminant would end up in the fast path

Could we cache a raw pointer to the contents?

Reply:
Alternatively, the structure is out small so I think having two almost-duplicates of it isn't the end in the world. In other locations (e.g. `EntityBuilder` vs. `EntityBuilderClone`) I've similarly favored duplication over wonky interfaces.

Reply:
I agree. Internal code duplication seems preferable to exposing internals. I modifying the last commit to take that route instead.